rslib/src/card_rendering/writer.rs

```diff
@@ -0,0 +1,210 @@
+                Node::SoundOrVideo(r) => self.write_sound(&mut buf, r),
+                Node::Tag(tag) => self.write_tag(&mut buf, tag),
+            };
+        }
+        buf
+    }
+
+    fn write_text(&mut self, buf: &mut String, txt: &str) {
+        buf.push_str(txt);
+    }
+
+    fn write_sound(&mut self, buf: &mut String, ressource: &str) {
+        write!(buf, "[sound:{}]", ressource).unwrap();
+    }
+
+    fn write_tag(&mut self, buf: &mut String, tag: &Tag) {
+        match tag {
+            Tag::Tts(tag) => self.write_tts_tag(buf, tag),
+            Tag::Other(tag) => self.write_other_tag(buf, tag),
+        };
+    }
+
+    fn write_tts_tag(&mut self, buf: &mut String, tag: &TtsTag) {
+        write!(buf, "[anki:tts").unwrap();
+
+        for (key, val) in [
+            ("lang", tag.lang),
+            ("voices", &tag.voices.join(",")),
+            ("speed", &tag.speed.to_string()),
+        ] {
+            self.write_tag_option(buf, key, val);
+        }
+        if let Some(blank) = tag.blank {
+            self.write_tag_option(buf, "cloze_blank", blank);
+        }
+        for (key, val) in &tag.options {
+            self.write_tag_option(buf, key, val);
+        }
+
+        write!(buf, "]{}[/anki:tts]", tag.content).unwrap();
+    }
+
+    fn write_other_tag(&mut self, buf: &mut String, tag: &OtherTag) {
+        write!(buf, "[anki:{}", tag.name).unwrap();
+        for (key, val) in &tag.options {
+            self.write_tag_option(buf, key, val);
+        }
+        buf.push(']');
+        self.write_tag_content(buf, tag.content);
+        write!(buf, "[/anki:{}]", tag.name).unwrap();
+    }
+
+    fn write_tag_option(&mut self, buf: &mut String, key: &str, val: &str) {
+        if val.contains::<&[char]>(&[']', ' ', '\t', '\r', '\n']) {
+            write!(buf, " {}=\"{}\"", key, val).unwrap();
+        } else {
+            write!(buf, " {}={}", key, val).unwrap();
+        }
+    }
+
+    fn write_tag_content(&mut self, buf: &mut String, content: &str) {
+        buf.push_str(content);
+    }
+}
+
+struct AvStripper;
+
+impl AvStripper {
+    fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Write for AvStripper {
+    fn write_sound(&mut self, _buf: &mut String, _ressource: &str) {}
+
+    fn write_tts_tag(&mut self, _buf: &mut String, _tag: &TtsTag) {}
+}
+
+struct AvExtractor<'a> {
+    side: char,
+    tags: Vec<pb::AvTag>,
+    tr: &'a I18n,
+}
+
+impl<'a> AvExtractor<'a> {
+    fn new(question_side: bool, tr: &'a I18n) -> Self {
+        Self {
+            side: if question_side { 'q' } else { 'a' },
+            tags: vec![],
+            tr,
+        }
+    }
+
+    fn write_play_tag(&self, buf: &mut String) {
+        write!(buf, "[anki:play:{}:{}]", self.side, self.tags.len()).unwrap();
+    }
+
+    fn transform_tts_content(&self, tag: &TtsTag) -> String {
+        strip_html_for_tts(tag.content).replace(
+            "[...]",
+            tag.blank.unwrap_or(&self.tr.card_templates_blank()),
```

Comment:
I'm surprised clippy didn't suggest using unwrap_or_else() here - maybe the reference made it think the call was cheap(er).

Reply:
Hmm, `.unwrap_or_else(|| &self.tr.card_templates_blank()),` wouldn't work though due to lifetimes.

Reply:
Ah yep. We could presumably work around it by converting to a string and dropping the reference, but if None is the common case, it's probably better left as is.